docs: Switch website from pkgdown to mkdocs-material with altdoc

etiennebacher opened the pull request:
Close #137 

See progress here: ~~https://etiennebacher.github.io/r-polars/site~~ No longer available (I removed the files so that this can be merged)

eitsupi commented:
Just a question: do we need to commit this large number of Markdown files?

etiennebacher commented:
All markdown files are used when we run `mkdocs build` (called by `altdoc::preview()`) so that it can build the `docs/site` folder. I'm still thinking of the best way to write a github action that would create all the docs automatically without having to commit anything other than in the usual R development workflow (like `pkgdown`).

etiennebacher commented:
But for now I commit everything because I put this branch on github pages so that we can have a live demo

eitsupi commented:
Thanks, it make sense now.

etiennebacher commented:
Note to self: we still need a `docs` folder that contains `mkdocs.yaml` and a subfolder `docs` containing custom CSS and logo because that's crucial to structure our docs. What we *don't* need in this folder are:

* the files for "Reference" because they can be generated automatically
* the vignettes because can be converted automatically -> caching them is not a good idea because even if the vignette in itself doesn't change, the functions used inside can change
* the `site` subfolder because it's generated by `mkdocs build`

etiennebacher commented:
Looks like the CI for documentation is ready, it only errors because I don't have the rights to create a branch on this repo

eitsupi commented:
Could the generated files be added to gitignore and then deleted?

etiennebacher commented:
I think the preview is still useful for now so if I delete the files we won't have it anymore. Do we do all the cleaning in R files here before merging?

sorhawell commented:
> Do we do all the cleaning in R files here before merging?

I didn't quite get this question. How can I help to bring this live :) what buttons should I press to give branch permission?

Currently pkgdown pushes to rpolars github.io with `secrets.DEPLOY_DOCS` token. Could you use that from the new workflow?

eitsupi commented on .github/workflows/documentation.yaml:
```diff
@@ -0,0 +1,109 @@
+name: mkdocs
+
+on:
+  push:
+    branches: main
```
This will always push from the main branch.
pkgdown automatically determines whether to update the root or the `dev` directory by the version number of the R package, but this trigger may be inappropriate if there is no such mechanism.

etiennebacher replied:
To me this means that the CI will run when there's a push on `main`, isn't that what we want? Also for now I don't think it's possible to have versioned documentation. I think `mkdocs` allows this but I didn't look into it and maybe it should be done in another PR later

eitsupi replied:
I don't think it is necessary to necessarily keep multiple versions of a document.

@sorhawell Do you have any idea which version of the document you would like to publish?
I think a while ago only tags would have triggered document updates.

sorhawell replied:
Before I only pushed docs for every binary release, but I think that was too rare. I think it makes sense to push docs at every push to main. At a later stage we could freeze the docs for each released version, similar to rust docs, but that is a luxury feature I think. Can a user who install polars from X commit build the docs herself locally?

etiennebacher replied:
Yes they would only need to run `altdoc::update_docs(custom_reference = "docs/make-docs.R")` (at the condition that they have everything needed to build the package).

But the [docs versioning](https://squidfunk.github.io/mkdocs-material/setup/setting-up-versioning/?h=versioni) is also a feature of `mkdocs`, we could enable it later

eitsupi replied:
I agree that it is not really necessary to keep the documentation for the latest release since this package is not yet a CRAN release.

> they would only need to run `altdoc::update_docs(custom_reference = "docs/make-docs.R")` (at the condition that they have everything needed to build the package).

Perhaps it would be worthwhile to manage jobs on this using Makefile, etc. I have created an issue #173

etiennebacher replied:
I never used Makefiles so I don't really understand what it would bring here. Is it to run `devtools::document` + `devtools::check` + `altdoc::update_docs(custom_reference = "docs/make-docs.R")` all at once? If so, how do you run it locally?

vincentarelbundock replied:
You just type `make` in a bash/zsh terminal. Once you are ready, just post different code blocks with the different workflows that we need (series of commands) in the Makefile Issue, and one of us (I can) will write one.

eitsupi commented on .github/workflows/documentation.yaml:
```diff
@@ -0,0 +1,109 @@
+on:
+  push:
+    branches: main
+  pull_request:
+    branches: main
```
I think we need to be careful not to push websites by PR.

etiennebacher replied:
Same as above, we want the CI to trigger on pull requests to see if a specific PR creates an error in the documentation. However, we shouldn't push the changes indeed so we need to add `github.event_name != 'pull_request'` at the end of the CI to avoid committing every time there's a PR

eitsupi commented on .github/workflows/documentation.yaml:
```diff
@@ -0,0 +1,109 @@
+        shell: Rscript {0}
+
+        # https://www.mkdocs.org/user-guide/deploying-your-docs/
+      - name: Build site and deploy to GitHub pages
+        run: cd docs && mkdocs gh-deploy --force
```
Note that we are not currently using the gh-pages branch of this repository.

https://github.com/pola-rs/r-polars/blob/2b04d3445dde103ff41099b91a7846eb4cc7f0af/.github/workflows/pkgdown.yaml#L107-L115

eitsupi commented on .github/workflows/documentation.yaml:
```diff
@@ -97,13 +100,28 @@ jobs:
+        if: github.event_name != 'pull_request'
+        uses: JamesIves/github-pages-deploy-action@v4
+        with:
+          clean: false
```
Perhaps if this is not set to `true`, the pkgdown files will continue to remain.

```suggestion
          clean: true
```

In this case, we also need to make sure that files like `.nojekyll` continue to remain.

etiennebacher commented:
> Currently [pkgdown pushes to rpolars github.io with](https://github.com/pola-rs/r-polars/blob/71e4e58d645ca5dcc11a9916803c70f90e1a481e/.github/workflows/pkgdown.yaml#L105) secrets.DEPLOY_DOCS token. Could you use that from the new workflow?

We now updated the workflow to also push to this repo. I removed most of the docs files (that will be automatically generated by the workflow) so once @eitsupi's comment is addressed I think we can merge this

eitsupi commented:
Thank you for tackling this tough task!
I would like to try a local build later and push some minor updates.

eitsupi commented:
I have made some minor updates.

- Rbuildignore and gitignore some files
  - Since `mkdocs.yml` is both the template and an auto generated file, CI was set up to copy the template `mkdocs.orig.yml` to `mkdocs.yml` just before executing `altdoc::update_docs()`.
    Perhaps it would be worth adding a new feature that could specify a template for `altdoc::update_docs()`
- To check the results of the CI run, I overwrote the workflow definition file with the existing pkgdown one.

If there are no problems with this change, it is working well and can be merged. Great job!

![image](https://user-images.githubusercontent.com/50911393/234012663-9145c6a6-6d0c-4694-be6c-82b47d1460b2.png)

eitsupi commented on docs/mkdocs.yml:
```diff
@@ -0,0 +1,520 @@
+nav:
+  - Home: README.md
+  - Get started: articles/polars.md
+  - Reference:
```
Thanks, as I mentioned in my comment (https://github.com/pola-rs/r-polars/pull/138#issuecomment-1520184506), I think it is better to separate the templates from the generated files, given that they may be updated automatically at build time.

etiennebacher commented:
Looks great, thanks for cleaning all of this @eitsupi, it's much better to keep a clear template like this. One thing: should we rename `pkgdown.yml` to something else?

Otherwise, if tests pass, I'm ok to merge

eitsupi commented:
> One thing: should we rename `pkgdown.yml` to something else?

My understanding is that CI will not run if renamed in this PR.
It would make sense to change this in a follow-up PR after merging this.

sorhawell commented:
Looking forward to see the new docs. pkgdown was a good stepping stone, but just deprecate it whenever ready :)

etiennebacher commented:
Looks like it worked! :tada: :tada: 

I'll make another PR soon with various tweaks

vincentarelbundock commented:
This is *amazing* work and a massive improvement over `pkgdown`. Thanks!

eitsupi commented:
Awesome!

(I forgot to remove the dev doc badge from README...)
